### PR TITLE
Backport of Quota in perc port to 0.2.3

### PR DESCRIPTION
**Pull request by hisundar**
Backport of https://github.com/CorfuDB/CorfuDB/pull/2063

### CODE REVIEW

File: test/src/test/java/org/corfudb/integration/AbstractIT.java [hunk 1]
```diff
@@ -379,7 +379,7 @@ public String getOptionsString() {
             }
 
             if (logSizeQuota != null) {
-                command.append(" --log-size-quota-bytes ").append(logSizeQuota);
+                command.append(" --log-size-quota-percentage ").append(logSizeQuota);
```
Comment (lixinchengdu) [inline test/src/test/java/org/corfudb/integration/AbstractIT.java]: logSizeQuota -> logSizeLimitPercentage

File: test/src/test/java/org/corfudb/infrastructure/ServerContextBuilder.java [hunk 2]
```diff
@@ -38,7 +38,7 @@
     String address = "test";
     int port = 9000;
     String seqCache = "1000";
-    String logSizeQuota = "-1";
+    String logSizeQuota = "100";
```
[the remark this answers is not in the document]
Comment (lixinchengdu) [reply test/src/test/java/org/corfudb/infrastructure/ServerContextBuilder.java]: The same name has been used for resource

File: test/src/test/java/org/corfudb/runtime/clients/LogUnitHandlerTest.java [hunk 3]
```diff
@@ -184,11 +187,14 @@ public void quotaExceededTest() throws Exception {
...
         final int maxLogSizeInBytes = 1000;
+        FileStore corfuDirBackend = Files.getFileStore(Paths.get(dirPath));
+        final long fsSize = corfuDirBackend.getTotalSpace();
+        final double maxLogSizeInPercentage = maxLogSizeInBytes * 100.0 / fsSize;
         ServerContext sc = serverContext = new ServerContextBuilder()
                 .setMemory(false)
                 .setLogPath(dirPath)
                 .setServerRouter(serverRouter)
-                .setLogSizeQuota(Long.toString(maxLogSizeInBytes))
+                .setLogSizeQuota(Double.toString(maxLogSizeInPercentage))
```
Comment (lixinchengdu) [inline test/src/test/java/org/corfudb/runtime/clients/LogUnitHandlerTest.java]: The same issue. logSizeQuota -> logSizeLimitPercentage

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 4]
```diff
@@ -120,15 +121,15 @@ public StreamLogFiles(ServerContext serverContext, boolean noVerify) {
         channelsToSync = new HashSet<>();
         this.verify = !noVerify;
         this.dataStore = StreamLogDataStore.builder().dataStore(serverContext.getDataStore()).build();
-        String logSizeQuotaParam = (String) serverContext.getServerConfig().get("--log-size-quota-bytes");
+        String logSizeQuotaParam = (String) serverContext.getServerConfig().get("--log-size-quota-percentage");
 
-        logSizeLimit = Long.parseLong(logSizeQuotaParam);
+        logSizeLimitPercentage = Double.parseDouble(logSizeQuotaParam);
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: rename logSizeQuotaParam to another name? The name of 'logSizeQuota' has been used for resource

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 5]
```diff
@@ -165,11 +166,12 @@ private void initStreamLogDirectory() {
                 throw new LogUnitException("Cannot start Corfu on a read-only filesystem:" + corfuDir);
             }
 
-            if (logSizeLimit != -1 && logSizeLimit > corfuDirBackend.getTotalSpace()) {
-                String msg = String.format("Invalid quota: quota(%d) is greater than the file store size (%d)",
-                        logSizeLimit, corfuDirBackend.getTotalSpace());
+            if (logSizeLimitPercentage < 0.0 || 100.0 < logSizeLimitPercentage) {
```
Comment (lixinchengdu) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Not sure whether we need to consider the precision of double comparison here. Generally speaking, when comparing double numbers, epsilon is usually used. 

https://stackoverflow.com/questions/356807/java-double-comparison-epsilon

File: test/src/test/java/org/corfudb/infrastructure/ServerContextBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hisundar) [reply test/src/test/java/org/corfudb/infrastructure/ServerContextBuilder.java]: done.

File: test/src/test/java/org/corfudb/infrastructure/ServerContextBuilder.java [hunk 6]
```diff
@@ -63,7 +63,7 @@ public ServerContext build() {
                 .put("--Threads", numThreads)
                 .put("--HandshakeTimeout", handshakeTimeout)
                 .put("--sequencer-cache-size", seqCache)
-                .put("--log-size-quota-bytes", logSizeQuota)
+                .put("--log-size-quota-percentage", logSizeQuota)
```
[the remark this answers is not in the document]
Comment (hisundar) [reply test/src/test/java/org/corfudb/infrastructure/ServerContextBuilder.java]: done.

File: test/src/test/java/org/corfudb/integration/AbstractIT.java (hunk 1 above)
Comment (hisundar) [reply test/src/test/java/org/corfudb/integration/AbstractIT.java]: done.

File: test/src/test/java/org/corfudb/runtime/clients/LogUnitHandlerTest.java (hunk 3 above)
Comment (hisundar) [reply test/src/test/java/org/corfudb/runtime/clients/LogUnitHandlerTest.java]: done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java (hunk 4 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java (hunk 5 above)
Comment (hisundar) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: We could, but here it is a simple boundary sanity check right, so do we need precision at all?

Comment (lixinchengdu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Yes. I share with your opinion. Sanity check does not need to be strict.

**Pull request merged**